core/src/services/supabase/core.rs

```diff
@@ -0,0 +1,222 @@
+    pub endpoint: String,
+
+    /// The key used for authorization
+    /// If loaded, the read operation will always access the nonpublic resources.
+    /// If you want to read the public resources, please do not set the key.
+    pub auth_key: Option<HeaderValue>,
+
+    pub http_client: HttpClient,
+}
+
+impl Debug for SupabaseCore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SupabaseCore")
+            .field("root", &self.root)
+            .field("bucket", &self.bucket)
+            .field("endpoint", &self.endpoint)
+            .finish_non_exhaustive()
+    }
+}
+
+impl SupabaseCore {
+    pub fn new(
+        root: &str,
+        bucket: &str,
+        endpoint: &str,
+        auth_key: Option<HeaderValue>,
+        client: HttpClient,
+    ) -> Self {
+        Self {
+            root: root.to_string(),
+            bucket: bucket.to_string(),
+            endpoint: endpoint.to_string(),
+            auth_key,
+            http_client: client,
+        }
+    }
+
+    /// Add authorization header to the request if the key is set. Otherwise leave
+    /// the request as-is.
+    pub fn sign<T>(&self, req: &mut Request<T>) -> Result<()> {
+        if let Some(k) = &self.auth_key {
+            let v = HeaderValue::from_str(&format!("Bearer {}", k.to_str().unwrap())).unwrap();
+            req.headers_mut().insert(http::header::AUTHORIZATION, v);
+        }
+        Ok(())
+    }
+}
+
+// requests
+impl SupabaseCore {
+    pub fn supabase_upload_object_request(
+        &self,
+        path: &str,
+        size: Option<usize>,
+        content_type: Option<&str>,
+        body: AsyncBody,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+        let url = format!(
+            "{}/storage/v1/object/{}/{}",
+            self.endpoint,
+            self.bucket,
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::post(&url);
+
+        if let Some(size) = size {
+            req = req.header(CONTENT_LENGTH, size)
+        }
+
+        if let Some(mime) = content_type {
+            req = req.header(CONTENT_TYPE, mime)
+        }
+
+        let req = req.body(body).map_err(new_request_build_error)?;
+
+        Ok(req)
+    }
+
+    pub fn supabase_get_object_public_request(&self, path: &str) -> Result<Request<AsyncBody>> {
+        self.supabase_get_object_request(path, false)
+    }
+
+    pub fn supabase_get_object_auth_request(&self, path: &str) -> Result<Request<AsyncBody>> {
+        self.supabase_get_object_request(path, true)
+    }
+
+    pub fn supabase_get_object_info_public_request(
+        &self,
+        path: &str,
+    ) -> Result<Request<AsyncBody>> {
+        self.supabase_get_object_info_request(path, false)
+    }
+
+    pub fn supabase_get_object_info_auth_request(&self, path: &str) -> Result<Request<AsyncBody>> {
+        self.supabase_get_object_info_request(path, true)
+    }
+}
```

Comment:
I feel like those logic is a bit overdesigned. `Core` doesn't need to expose those internal logic to `Backend`.

Reply:
Fixed